### PR TITLE
Deploy (& dispose) E2E test apps in parallel (Take 3)

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
Third take on #5660

Should make E2E tests noticeably faster by deploying (and disposing) apps in parallel.

This should have an especially large benefit on tests that deploy many apps (like service_invocation) and/or on AKS, where we need to wait for an Azure LoadBalancer to be created for each Service.

### CODE REVIEW

File: tests/runner/testresource.go [hunk 1]
```diff
@@ -90,26 +90,60 @@ func (r *TestResources) FindActiveResource(name string) Disposable {
...
+	}
+
+	for i := 0; i < resourceCount; i++ {
+		err := <-errs
 		if err != nil {
 			return err
 		}
 	}
+
 	return nil
```
Comment (JoshVanL) [inline tests/runner/testresource.go:116]: We should `errors.Join` here, instead of returning on the first error.

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: tests/runner/testresource.go (hunk 1 above)
Comment (ItalyPaleAle) [reply tests/runner/testresource.go:116]: Done

File: tests/runner/testresource.go [hunk 2]
```diff
@@ -90,26 +90,60 @@ func (r *TestResources) FindActiveResource(name string) Disposable {
...
+	}
+
+	for i := 0; i < resourceCount; i++ {
+		err := <-errs
 		if err != nil {
+			os.Stderr.WriteString(err.Error() + "\n")
 			retErr = err
-			fmt.Fprintf(os.Stderr, "Failed to tear down %s. got: %q", dr.Name(), err)
 		}
 	}
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/runner/testresource.go:145]: Done

**Comment by JoshVanL** [comment]
LGTM

**Comment by mukundansundar** [comment]
> # Dapr E2E test
> 🔗 **[Link to Action run](https://github.com/dapr/dapr/actions/runs/5278724994)**
> 
> Commit ref: [87c458f](https://github.com/dapr/dapr/commit/87c458fdd157805e3aa5036e6ff93f602de39c63)
> 
> ## ✅ Build succeeded for linux/amd64
> * Image tag: `dapre2edc36b0bf99l`
> * Test image tag: `dapre2edc36b0bf99l`
> 
> ## ✅ Infrastructure deployed
> Cluster	Resource group name	Azure region
> Linux	`Dapr-E2E-dapre2edc36b0bf99l`	westus3
> Windows	`Dapr-E2E-dapre2edc36b0bf99w`	westus3
> Linux/arm64	`Dapr-E2E-dapre2edc36b0bf99la`	eastus
> ## ✅ Build succeeded for windows/amd64
> * Image tag: `dapre2edc36b0bf99w`
> * Test image tag: `dapre2edc36b0bf99w`
> 
> ## ✅ Tests succeeded on linux/amd64
> * Image tag: `dapre2edc36b0bf99l`
> * Test image tag: `dapre2edc36b0bf99l`
> 
> ## ✅ Tests succeeded on windows/amd64
> * Image tag: `dapre2edc36b0bf99w`
> * Test image tag: `dapre2edc36b0bf99w`

@ItalyPaleAle I am seeing an increase in the time it takes to complete the E2E ... Existing code successful run https://github.com/dapr/dapr/actions/runs/5274418893 vs the run above ~30 minutes more. 


I believe this might be due to the timeouts that have been increased/doubled. 
Is it needed to increase the timeout?

**Comment by ItalyPaleAle** [comment]
@mukundansundar it seems very variable. This test you linked did take a long time, but the previous one took 30 mins less: https://github.com/dapr/dapr/actions/runs/5205873623

No idea how to interpret this 🤷 Sometimes it takes longer in the build and deploy steps, but in the test you linked, the tests themselves took longer.

The increased timeouts are just for API calls to the K8s control plane, and they're still only a few seconds longer. I doubt that's what's causing issues...

File: tests/runner/testresource.go [hunk 3]
```diff
@@ -90,29 +91,65 @@ func (r *TestResources) FindActiveResource(name string) Disposable {
...
+	for i := 0; i < resourceCount; i++ {
+		err := <-errs
 		if err != nil {
-			return err
+			allErrs = append(allErrs, err)
 		}
 	}
-	return nil
+
+	return errors.Join(allErrs...)
```
Comment (mukundansundar) [inline tests/runner/testresource.go:118]: ideally we can use waitgroup here

File: tests/runner/testresource.go [hunk 4]
```diff
@@ -90,29 +91,65 @@ func (r *TestResources) FindActiveResource(name string) Disposable {
...
 }
 
 // TearDown initializes the resources by calling Dispose.
-func (r *TestResources) tearDown() (retErr error) {
-	retErr = nil
-	for dr := r.popActiveResource(); dr != nil; dr = r.popActiveResource() {
-		err := dr.Dispose(false)
+func (r *TestResources) tearDown() error {
+	resourceCount := 0
```
Comment (mukundansundar) [inline tests/runner/testresource.go:123]: same here...

File: tests/runner/testresource.go (hunk 3 above)
Comment (ItalyPaleAle) [reply tests/runner/testresource.go:118]: With waitgroup it's harder to collect errors (without causing race conditions for concurrent access to a slice). This does achieve a similar effect while also allowing the collection of errors in a thread-safe way.

**Review by mukundansundar — APPROVED** [review]

**Pull request merged**